testsuite/integration/basic/src/test/java/org/wildfly/test/integration/microprofile/metrics/application/MicroProfileMetricsDifferentFormatsValueTestCase.java

```diff
@@ -0,0 +1,170 @@
+     * ManagementClient provides management port and hostname
+     */
+    @ContainerResource
+    ManagementClient managementClient;
+
+    /**
+     * Check Metered annotation
+     */
+    @Test
+    public void testMetered(@ArquillianResource URL url) throws Exception {
+        // requests
+        performCallsForOneEndPoint(url, "metered");
+
+        // asserts
+        commonAsserts("metered");
+        diffAssert("metered", "count", 1, "metered_total", 0.0);
+    }
+
+    /**
+     * Check Timed annotation
+     */
+    @Test
+    public void testTimed(@ArquillianResource URL url) throws Exception {
+        // requests
+        performCallsForOneEndPoint(url, "timed");
+
+        // asserts for same values as metered
+        diffAssert("timed", "count", 1, "timed_seconds_count", 0.0);
+        commonAsserts("timed");
+        // followed values has different units for json and prometheus format. Minutes is unit for json, seconds is unit for prometheus.
+        diffAssert("timed", "min", 60, "timed_min_seconds", 1.0E-10 * TimeoutUtil.adjust(1));
+        diffAssert("timed", "mean", 60, "timed_mean_seconds", 1.0E-10 * TimeoutUtil.adjust(1));
+        diffAssert("timed", "max", 60, "timed_max_seconds", 1.0E-10 * TimeoutUtil.adjust(1));
+        diffAssert("timed", "stddev", 60, "timed_stddev_seconds", 1.0E-10 * TimeoutUtil.adjust(1));
+        diffAssert("timed", "p50", 60, "timed_seconds{quantile=\"0.5\"}", 1.0E-10 * TimeoutUtil.adjust(1));
+        diffAssert("timed", "p75", 60, "timed_seconds{quantile=\"0.75\"}", 1.0E-10 * TimeoutUtil.adjust(1));
+        diffAssert("timed", "p95", 60, "timed_seconds{quantile=\"0.95\"}", 1.0E-10 * TimeoutUtil.adjust(1));
+        diffAssert("timed", "p98", 60, "timed_seconds{quantile=\"0.98\"}", 1.0E-10 * TimeoutUtil.adjust(1));
+        diffAssert("timed", "p99", 60, "timed_seconds{quantile=\"0.99\"}", 1.0E-10 * TimeoutUtil.adjust(1));
+        diffAssert("timed", "p999", 60, "timed_seconds{quantile=\"0.999\"}", 1.0E-10 * TimeoutUtil.adjust(1));
+    }
+
+    /**
+     * Common asserts for both timed and metered annotations
+     */
+    private void commonAsserts(String measuredName) {
+        diffAssert(measuredName, "meanRate", 1, measuredName + "_rate_per_second", TimeoutUtil.adjust(15));
+        diffAssert(measuredName, "oneMinRate", 1, measuredName + "_one_min_rate_per_second", 0.1 * TimeoutUtil.adjust(1));
+        diffAssert(measuredName, "fiveMinRate", 1, measuredName + "_five_min_rate_per_second", 0.1 * TimeoutUtil.adjust(1));
+        diffAssert(measuredName, "fifteenMinRate", 1, measuredName + "_fifteen_min_rate_per_second", 0.1 * TimeoutUtil.adjust(1));
+    }
+
+    /**
+     * Assert maximal diff for json and prometheus values:
+     * Assert.true(application:{prometheusName} - ({jsonName}->{jsonSubName} * jsonMultiplier) <= {maximalDiff})
+     */
+    private void diffAssert(String jsonName, String jsonSubName, int jsonMultiplier, String prometheusName, double maximalDiff) {
```

Comment:
I don't think that these differences from JSON and Prometheus output should be used to test metrics.
There is no guarantee that quick successive calls to the JSON & Prometheus output returns "identical" figures.

I'd prefer that assertions are tested separately on Prometheus and JSON output independently based on different in application invocations (that can be controlled).

Reply:
@jmesnil:

> There is no guarantee that quick successive calls to the JSON & Prometheus output returns "identical" figures.

There will be never identical, because [json metrics are gotten some time after Prometheus metrics](https://github.com/wildfly/wildfly/pull/12036/files#diff-bcef9e5450c8c46226465018472184c6R170).

But quick successive calls won't be never so quick, because of [this Thread.sleep(100)](https://github.com/wildfly/wildfly/pull/12036/files#diff-bcef9e5450c8c46226465018472184c6R156)

These asserts are not assertEquals, asserts have maximalDiff, so these asserts should pass. These asserts should be OK on weaker machines thanks to TimeoutUtil.

> I'd prefer that assertions are tested separately on Prometheus and JSON output independently based on different in application invocations (that can be controlled).

We still can't say the output value precisely as well in this approach. Test depends on env, [10 requests from the test](https://github.com/wildfly/wildfly/pull/12036/files#diff-bcef9e5450c8c46226465018472184c6R160) takes different time on different machines. Anyway yes, this approach is also an option. But I prefer to compare json and prometheus values, because these values should be similar (because prometheus and json requests should take similar time amout on one machine). This test approach also check the original [#39](https://github.com/smallrye/smallrye-metrics/issues/39) and [#40](https://github.com/smallrye/smallrye-metrics/issues/40) bugs better. WDYT?

Reply:
I don't like this at all. This kind of detail should be tested in Smallrye.  I can understand some degree of testing of smallrye details within the WF testsuite, but this is class is too low level.